db_master_deploy/deploy.sh

```diff
@@ -0,0 +1,331 @@
+    array=(${source_object//./ })
+    # check source objects
+    if [ "${#array[@]}" -ne "3" -a "${#array[@]}" -ne "1" ]; then
+        echo "table data sources have to be formatted like this: db.schema.table, database sources like this: db" >&2
+        exit 1
+    fi
+    # tables go here
+    if [ "${#array[@]}" -eq "3" ]; then
+        echo "processing table $source_object..."
+        source_db=${array[0]}
+        source_schema=${array[1]}
+        source_table=${array[2]}
+        target_db=${source_db%_*}"_"$target # lubis_3d_master -> lubis_3d_$target
+        target_schema="$source_schema"
+        target_table="$source_table"
+        source_id="$source_db.$source_schema.$source_table"
+        target_id="$target_db.$target_schema.$target_table"
+        array_target_table+=($target_id)
+        array_source_table+=($source_object)
+        array_target_combined+=($target_id)
+        check_table
+        check_source
+        copy_table
+    fi
+    # databases go here
+    if [ "${#array[@]}" -eq "1" ]; then
+        echo "processing database $source_object..."        
+        source_db=${array[0]}
+        target_db=${source_db%_*}"_"$target
+        target_db_tmp=$target_db""_tmp_deploy
+        array_target_db+=($target_db)
+        array_source_db+=($source_object)
+        array_target_combined+=($target_db)
+        check_database
+        check_source
+        bod_create_archive
+        copy_database        
+    fi    
+done
+
+END=$(date +%s%3N)
+echo "master has been updated in $((END-START)) milliseconds"
+
+# wait for all slaves
+while :
+do
+    diff=999
+    read slaves diff <<< $(psql \
+    -X \
+    -U pgkogis \
+    -h localhost \
+    -d postgres \
+    --single-transaction \
+    --set ON_ERROR_STOP=on \
+    --no-align \
+    -t \
+    --field-separator ' ' \
+    --quiet \
+    -c "select count(1) as slaves, coalesce(sum(pg_xlog_location_diff(sent_location,replay_location)),0) as diff FROM pg_stat_replication where state IN('streaming') ;")
+
+    # default value if connection above does not work
+    FOO=${diff:=999}
+
+    if [[ $diff -eq 0  && $slaves -eq $attached_slaves ]]
+    then
+        END_slaves=$(date +%s%3N)
+        echo "$slaves slaves have been updated in $((END_slaves-END)) milliseconds"
+        break
+    fi
+done
+echo "finished ${0##*/} $* in $((END_slaves-START)) milliseconds"
+
+# concatenate arrays for dml and ddl trigger
+source_db=$(IFS=, ; echo "${array_source_db[*]}")
+target_combined=$(IFS=, ; echo "${array_target_combined[*]}")
+
+# fire dml trigger in sub shell
+# redirect customized stdout and stderr to standard ones
+(
+bash "$MY_DIR/dml_trigger.sh" -s $target_combined -t $target 1>&5 2>&6
+)
+
+if [ "${#array_target_db[@]}" -gt "0" ]; then
+    # fire ddl trigger in sub shell
+    # redirect customized stdout and stderr to standard ones    
+    (
+    bash "$MY_DIR/ddl_trigger.sh" -s $source_db -t $target 1>&5 2>&6
+    )  
+fi
+
+rm -f "$lockfile"
```

Comment:
Thanks for the clarification. I think this should be documented somewhere. People should know what to do when the script fails. And they should know what to look for in the output.

Long-term goal should be to have a streamlined deploy process, managed by jenkins (using somethin like the jenkins build pipeline [1]). You configure what you want to deploy (basically, what is in the to-deploy doc) and then let it run with human input only when necessary/needed (in case of failure, in case where manual tests are needed, in case to confirm to continue, etc). We are far away from this, but should work in this direction. One requirement for this is that each sup-step of a deploy should be revertable.

[1] https://wiki.jenkins-ci.org/display/JENKINS/Build+Pipeline+Plugin